File: apps/php/design-pattern/tests/InterpreterTest.php
<?php

declare(strict_types=1);

namespace DesignPattern\Tests;

use DesignPattern\Interpreter\All;
use DesignPattern\Interpreter\AndExpression;
use DesignPattern\Interpreter\Bigger;
use DesignPattern\Interpreter\FileName;
use DesignPattern\Interpreter\NotExpression;
use DesignPattern\Interpreter\OrExpression;
use PHPUnit\Framework\TestCase;

class InterpreterTest extends TestCase
{
    private string $testDir;

    protected function setUp(): void
    {
        $this->testDir = sys_get_temp_dir() . '/design_pattern_interpreter_test_' . uniqid();
        mkdir($this->testDir, 0777, true);

        file_put_contents($this->testDir . '/small.txt', 'hi');
        file_put_contents($this->testDir . '/big.txt', str_repeat('x', 200));
        file_put_contents($this->testDir . '/image.png', str_repeat('y', 300));
        file_put_contents($this->testDir . '/doc.txt', str_repeat('z', 50));
    }

    protected function tearDown(): void
    {
        $files = glob($this->testDir . '/*');
        if ($files) {
            array_map('unlink', $files);
        }
        if (is_dir($this->testDir)) {
            rmdir($this->testDir);
        }
    }

    public function testAllExpression(): void
    {
        $expr = new All();
        $result = $expr->evaluate($this->testDir);

        $this->assertCount(4, $result);
    }

    public function testFileNameExpression(): void
    {
        $expr = new FileName('*.txt');
        $result = $expr->evaluate($this->testDir);

        $this->assertCount(3, $result);
        foreach ($result as $path) {
            $this->assertStringEndsWith('.txt', $path);
        }
    }

    public function testBiggerExpression(): void
    {
        $expr = new Bigger(100);
        $result = $expr->evaluate($this->testDir);

        $this->assertCount(2, $result);
    }

    public function testAndExpression(): void
    {
        $expr = new AndExpression(new FileName('*.txt'), new Bigger(100));
        $result = $expr->evaluate($this->testDir);

        $this->assertCount(1, $result);
        $this->assertStringContainsString('big.txt', $result[0]);
    }

    public function testOrExpression(): void
    {
        $expr = new OrExpression(new FileName('*.png'), new Bigger(100));
        $result = $expr->evaluate($this->testDir);

        // image.png (matches both) + big.txt (matches bigger)
        $this->assertCount(2, $result);
    }

    public function testNotExpression(): void
    {
        $expr = new NotExpression(new FileName('*.txt'));
        $result = $expr->evaluate($this->testDir);

        $this->assertCount(1, $result);
        $this->assertStringContainsString('image.png', $result[0]);
    }

    public function testComplexExpression(): void
    {
        // txt files that are NOT bigger than 100 bytes
        $expr = new AndExpression(
            new FileName('*.txt'),
            new NotExpression(new Bigger(100))
        );
        $result = $expr->evaluate($this->testDir);

        $names = array_map('basename', $result);
        $this->assertContains('small.txt', $names);
        $this->assertContains('doc.txt', $names);
        $this->assertNotContains('big.txt', $names);
    }
}

File: apps/ruby/design-pattern/16-interpreter/test/interpreter_test.rb
require_relative "test_helper"

class InterpreterTest < Minitest::Test
  def setup
    @test_dir = Dir.mktmpdir("interpreter_test")
    File.write(File.join(@test_dir, "big.txt"), "x" * 200)
    File.write(File.join(@test_dir, "small.txt"), "x" * 5)
    File.write(File.join(@test_dir, "data.csv"), "a,b,c\n1,2,3\n")
    File.write(File.join(@test_dir, "readme.md"), "# README")
  end

  def teardown
    FileUtils.rm_rf(@test_dir)
  end

  def test_all_returns_every_file
    result = All.new.evaluate(@test_dir)
    assert_equal 4, result.size
  end

  def test_filename_matches_pattern
    result = FileName.new("*.txt").evaluate(@test_dir)
    basenames = result.map { |file| File.basename(file) }

    assert_includes basenames, "big.txt"
    assert_includes basenames, "small.txt"
    refute_includes basenames, "data.csv"
  end

  def test_and_combines_expressions
    expr = And.new(FileName.new("*.txt"), Bigger.new(100))
    result = expr.evaluate(@test_dir)
    basenames = result.map { |file| File.basename(file) }

    assert_includes basenames, "big.txt"
    refute_includes basenames, "small.txt"
  end

  def test_not_negates_expression
    expr = Not.new(FileName.new("*.txt"))
    result = expr.evaluate(@test_dir)
    basenames = result.map { |file| File.basename(file) }

    assert_includes basenames, "data.csv"
    refute_includes basenames, "big.txt"
  end
end
